Gems/MotionMatching/Code/Source/FeatureAngularVelocity.cpp

```diff
@@ -0,0 +1,158 @@
+
+            const AZ::Vector3& angularVelocity = velocityPoseData->GetAngularVelocities()[m_jointIndex];
+            context.m_featureMatrix.SetVector3(context.m_frameIndex, m_featureColumnOffset, angularVelocity);
+        }
+        context.m_posePool.FreePose(tempPose);
+    }
+
+    void FeatureAngularVelocity::FillQueryVector(QueryVector& queryVector, const QueryVectorContext& context)
+    {
+        PoseDataJointVelocities* velocityPoseData = context.m_currentPose.GetPoseData<PoseDataJointVelocities>();
+        AZ_Assert(velocityPoseData, "Cannot calculate velocity feature cost without joint velocity pose data.");
+        const AZ::Vector3 currentVelocity = velocityPoseData->GetAngularVelocity(m_jointIndex);
+
+        queryVector.SetVector3(currentVelocity, m_featureColumnOffset);
+    }
+
+    float FeatureAngularVelocity::CalculateFrameCost(size_t frameIndex, const FrameCostContext& context) const
+    {
+        const AZ::Vector3 queryVelocity = context.m_queryVector.GetVector3(m_featureColumnOffset);
+        const AZ::Vector3 frameVelocity = context.m_featureMatrix.GetVector3(frameIndex, m_featureColumnOffset);
+
+        return CalcResidual(queryVelocity, frameVelocity);
+    }
+
+    void FeatureAngularVelocity::DebugDraw(
+        AzFramework::DebugDisplayRequests& debugDisplay,
+        const Pose& pose,
+        const AZ::Vector3& velocity,
+        size_t jointIndex,
+        size_t relativeToJointIndex,
+        const AZ::Color& color)
+    {
+        const Transform jointModelTM = pose.GetModelSpaceTransform(jointIndex);
+        const Transform relativeToWorldTM = pose.GetWorldSpaceTransform(relativeToJointIndex);
+
+        const AZ::Vector3 jointPosition = relativeToWorldTM.TransformPoint(jointModelTM.m_position);
+        const AZ::Vector3 velocityWorldSpace = relativeToWorldTM.TransformVector(velocity);
+
+        DebugDrawVelocity(debugDisplay, jointPosition, velocityWorldSpace * mm_debugDrawVelocityScale, color);
+    }
+
+    void FeatureAngularVelocity::DebugDraw(
+        AzFramework::DebugDisplayRequests& debugDisplay,
+        const Pose& currentPose,
+        const FeatureMatrix& featureMatrix,
+        const FeatureMatrixTransformer* featureTransformer,
+        size_t frameIndex)
+    {
+        if (m_jointIndex == InvalidIndex)
+        {
+            return;
+        }
+
+        AZ::Vector3 angularVelocity = featureMatrix.GetVector3(frameIndex, m_featureColumnOffset);
+        if (featureTransformer)
+        {
+            angularVelocity = featureTransformer->InverseTransform(angularVelocity, m_featureColumnOffset);
+        }
+        DebugDraw(debugDisplay, currentPose, angularVelocity, m_jointIndex, m_relativeToNodeIndex, m_debugColor);
+    }
+
+    void FeatureAngularVelocity::Reflect(AZ::ReflectContext* context)
+    {
+        AZ::SerializeContext* serializeContext = azrtti_cast<AZ::SerializeContext*>(context);
```

Comment:
uber nit: prefer `auto` here to avoid duplicated type info on the left and right of the expression

Reply:
Changed, will come with the next PR.